### PR TITLE
[Feat] Add support for pushing the amplification results in a MongoDB database

### PR DESCRIPTION
**Pull request by Tailp**
Hi :), original PR (#850) . This contains only the structure of the soon-to-be implemented. Awaiting for review @monperrus  @danglotb .

### CODE REVIEW

File: dspot/src/main/java/eu/stamp_project/dspot/selector/JacocoCoverageSelector.java [hunk 1]
```diff
@@ -222,6 +223,12 @@ public TestSelectorElementReport report() {
                     .append("%")
                     .append(AmplificationHelper.LINE_SEPARATOR);
             lastReport = new TestSelectorElementReportImpl(report.toString(), jsonReport(coverageResults));
+
+            // Report to mongodb
+            MongodbManager.getInstance().reportJacocotInformation(this.currentClassTestToBeAmplified.getQualifiedName(),
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/dspot/selector/JacocoCoverageSelector.java]: global state + mongo hardcoded to be avoided. All calls to `MongodbManager.getInstance()` should be replaced `config.getInformationCollector()`

**Comment by Tailp** [comment]
Hi @monperrus :) , I have attempted to resolved it and wait for new comments .

File: dspot/src/main/java/eu/stamp_project/dspot/selector/JacocoCoverageSelector.java [hunk 2]
```diff
@@ -222,6 +223,12 @@ public TestSelectorElementReport report() {
                     .append("%")
                     .append(AmplificationHelper.LINE_SEPARATOR);
             lastReport = new TestSelectorElementReportImpl(report.toString(), jsonReport(coverageResults));
+
+            // Report to mongodb
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/dspot/selector/JacocoCoverageSelector.java]: remove comment, it's not about Mongo anymore

File: dspot/src/main/java/eu/stamp_project/dspot/mongodb/DspotInformationCollector.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package eu.stamp_project.mongodb;
+
+import com.martiansoftware.jsap.JSAPResult;
+
+public interface DspotInformationCollector {
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/dspot/mongodb/DspotInformationCollector.java]: interface are essential to be documented. add one sentence of doc for the class and per method, ideally with an example

File: dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java [hunk 4]
```diff
@@ -226,6 +227,10 @@ public TestSelectorElementReport report() {
         }
         final String reportStdout = reportStdout();
         final TestClassJSON testClassJSON = reportJSONMutants();
+        // Also report to Mongodb
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java]: -comment

File: dspot/src/main/java/eu/stamp_project/utils/DSpotUtils.java [hunk 5]
```diff
@@ -134,6 +136,9 @@ public static void printAndCompileToCheck(CtType<?> type, File directory) {
                 LOGGER.warn("DSpot outputs it using full qualified names.");
                 LOGGER.warn("These problems can come from the fact your project use generated codes, such as Lombok annotations.");
                 printCtTypeToGivenDirectory(type, directory, false); //FIXME: analyse for optimisation (13% total execution time)
+            } else {
+                /*If test was able to compile, report to mongo*/
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/utils/DSpotUtils.java]: comment

File: dspot/src/main/java/eu/stamp_project/utils/options/Configuration.java [hunk 6]
```diff
@@ -23,6 +25,7 @@
 public class Configuration {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(Configuration.class);
+    private static final DspotInformationCollector mongodbManager = new MongodbManager();
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/utils/options/Configuration.java]: by default it should be a NullCollector (an empty implementation that does nothing)

File: dspot/src/main/java/eu/stamp_project/utils/options/JSAPOptions.java [hunk 7]
```diff
@@ -115,6 +116,8 @@ public static void parse(String[] args) {
                 testCases,
                 fullClasspath
         );
+        DspotInformationCollector mongodbManager = Configuration.getInformationCollector();
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/utils/options/JSAPOptions.java]: variable should be called `collector`

**Comment by monperrus** [comment]
@danglotb we're almost there. WDYT?

**Comment by danglotb** [comment]
Hello, the changes look good to me.

A question: In which format will you push the information in the database?
Could we rely on a single `String`?

If so, it might be better to have a single method to report the result of test selector, such as `reportTestSelectionInformation` rather than having multiples methods for each test selector, which might be tedious to maintain. WDYT?

**Comment by Tailp** [comment]
Hi @danglotb , It's in JSON format :) . A single string would need a bit of work to divide it to relevant parts before submitting for each database . How would the String look like ? :) .

File: dspot/src/main/java/eu/stamp_project/utils/DSpotUtils.java (hunk 5 above)
Comment (Tailp) [reply dspot/src/main/java/eu/stamp_project/utils/DSpotUtils.java]: I assume here you mean -comment :) .

**Comment by danglotb** [comment]
It can be the string representation of the JSON then, isn't it?

The test selector would create this JSON with its info, such as it is already done, _e.g._ in [PitMutantScoreSelector](https://github.com/STAMP-project/dspot/blob/master/dspot/src/main/java/eu/stamp_project/dspot/selector/PitMutantScoreSelector.java#L228), and give the string representation to the collector using the API `reportTestSelectionInformation(String s)`. 

We could also improve the [`eu.stamp_project.utils.report.output.selector.TestClassJSON`](https://github.com/STAMP-project/dspot/blob/master/dspot/src/main/java/eu/stamp_project/utils/report/output/selector/TestClassJSON.java) interface, that is implemented by all the `TestClassJSONs` classes, such as `eu.stamp_project.utils.report.output.selector.mutant.json.TestClassJSON` or `eu.stamp_project.utils.report.output.selector.coverage.json.TestClassJSON`.

Then, `reportTestSelectionInformation(String s)` would just transform the `String` into a correct JSON and push it on the DB (or push directly the JSON, depending on the paragraph above).

WDYT?

**Comment by Tailp** [comment]
@danglotb hi :) , yes we can have that . My thoughts, we can have a document taking the saved information in the previous step in the function reportJSONMutants so maybe reportTestSelectionInformation(String s) can be reportTestSelectionInformation() and it will send information in the implementation over to Mongodb. The document differs depending on which selector it was .

**Comment by Tailp** [comment]
@danglotb @monperrus resolved :) . Waiting for new comments.

**Comment by monperrus** [comment]
OK for me, thanks @Tailp. @danglotb could you do the last checks and merge? Thanks!

**Comment by danglotb** [comment]
I relaunched the build on Travis.

I dislike the fact that the JSON classes are using the collector.

For me, the JSON classes are pure data representation and the only available API should be the modification of these data.

I would rather prefer that the collector should be the one that uses the data. But it a conceptual vision of the classes' role.

Questions: 
- Could we implement a collector that pushes the data into a file? Could write some unit tests about this?
- Do you plan to implement the configuration through the command line in a future PR?

Thank you!

**Comment by Tailp** [comment]
@danglotb hi :), should I revert back to the previous structure proposed by @monperrus  or do you prefer to keep this structure ? . The previous structure does not embed code into the JSON classes, which seems to be a better alternative according to your comments. 

Pushing it to a file then read it would be a bit round about, while we can just give the infomation directly to the mongodb. 

which configuration do you mean ? :) .

**Comment by monperrus** [comment]
> Could we implement a collector that pushes the data into a file? Could write some unit tests about this?

Yes. Maybe in a subsequent PR?

> Do you plan to implement the configuration through the command line in a future PR?

That's an option.


What's the concrete next step to make progress here towards merging this baby PR?

**Comment by danglotb** [comment]
Hello, I would like to have the TestClassJSON to be pure data classes, _i.e._ without any logic such as the call to the `Configuration.getInformationCollector().reportSelectorInformation(report);`.

Concretely, changes `reportTestSelectionInformation` into a likewise `toString`. You can probably use `GSON` API to format correctly the data into JSON String style rather than by hands.
Puts the logic in the TestSelector, _i.e._ replaces
`testClassJSON.reportTestSelectionInformation();` 
by something like
`Configuration.getInformationCollector().reportSelectorInformation(testClassJSON.toString());`

And then, the `testClassJSON` contains only information and from the point the view of the system, it represents a JSON file.

On my side, I must fix Travis, it seems that it is able to download a dependency: 

```txt
[WARNING] Could not transfer metadata org.eclipse.platform:org.eclipse.osgi/maven-metadata.xml from/to sonatype-apache (https://repository.apache.org/releases/): Connect to repository.apache.org:443 [repository.apache.org/207.244.88.140] failed: Connection timed out (Connection timed out)
```

It stucks at this step, trying to download it, again and again, making the build fail.

**Comment by Tailp** [comment]
@danglotb Sure :) , it's resolved . Please have a look .

**Comment by danglotb** [comment]
Hello @Tailp.

Good for me, unless that the `toString()` method should not be declared in the interface and you should add an `@Override` annotation since you are overriding the `toString()` from `java.lang.Object`.

Still have some issues with Travis, dunno how to fix this yet. Sorry.

Thank you!

**Comment by Tailp** [comment]
@danglotb I understand, no worries :) . The toString() method is now removed from the interface.

**Comment by Tailp** [comment]
@danglotb hi :) , I got this error with Coverall
```
ERROR] Failed to execute goal org.eluder.coveralls:coveralls-maven-plugin:4.3.0:report (default-cli) on project dspot: I/O operation failed: No source found for eu/stamp_project/mongodb/NullCollector.java -> [Help 1]

[ERROR] 

[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.

[ERROR] Re-run Maven using the -X switch to enable full debug logging.

[ERROR] 

[ERROR] For more information about the errors and possible solutions, please read the following articles:

[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException

The command "./.travis/$SCRIPT $DSPOT_VERSION" exited with 1.
```
It does not make any sense to me since I have not used Coverall before. This also occured before with MongodbCollector.java but was somehow fixed after your fix :) . Maybe you happens to know the solution to this ? . Thanks .

**Comment by danglotb** [comment]
Hello @Tailp 

I'm looking at this.

**Comment by danglotb** [comment]
It is because the used package is `package eu.stamp_project.mongodb;` while your sources file are in `eu/stamp_project/dspot/mongodb`.

Please move your sources file, _i.e._ all the classes of the package `package eu.stamp_project.mongodb;` to the correct directory.

In addition to this, you can add this package to the package `eu.stamp_project.utils` , and so having the package `eu.stamp_project.utils.mongodb` in the directory `dspot/src/main/java/eu/stamp_project/utils/mongodb`.

You may also want put all the generic class into a package that is more generic, such as `eu.stamp_project.utils.collector` and put the mongodb specific classes into `eu.stamp_project.utils.collector.mongodb` in order to have a better hierarchy of the packages/files.

**Comment by Tailp** [comment]
> It is because the used package is `package eu.stamp_project.mongodb;` while your sources file are in `eu/stamp_project/dspot/mongodb`.
> 
> Please move your sources file, _i.e._ all the classes of the package `package eu.stamp_project.mongodb;` to the correct directory.
> 
> In addition to this, you can add this package to the package `eu.stamp_project.utils` , and so having the package `eu.stamp_project.utils.mongodb` in the directory `dspot/src/main/java/eu/stamp_project/utils/mongodb`.
> 
> You may also want put all the generic class into a package that is more generic, such as `eu.stamp_project.utils.collector` and put the mongodb specific classes into `eu.stamp_project.utils.collector.mongodb` in order to have a better hierarchy of the packages/files.

Thank you very much :) . I'll try fixing it now once again .

**Comment by danglotb** [comment]
Ready to merge for me. @Tailp All good?

**Comment by Tailp** [comment]
@danglotb yes :) .

**Pull request merged**